engine/buildkit/filesync.go

```diff
@@ -0,0 +1,488 @@
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get client metadata: %w", err)
+	}
+
+	// for now, require that the requester is the owner of the session, i.e. you
+	// can only export to yourself, not to others
+	if clientMetadata.ClientID != opts.DestClientID {
+		return nil, nil, errors.New("local dir export requester is not the owner of the dest session")
+	}
+
+	if opts.Path == "" {
+		return nil, nil, fmt.Errorf("missing local dir export path")
+	}
+
+	sess, err := c.SessionManager.Get(stream.Context(), opts.DestClientID, false)
+	if err != nil {
+		return nil, nil, err
+	}
+	md := opts.ToGRPCMD()
+	ctx := metadata.NewIncomingContext(stream.Context(), md) // TODO: needed too?
+	ctx = metadata.NewOutgoingContext(ctx, md)
+	return ctx, &localExportOpts{
+		LocalExportOpts: opts,
+		session:         sess,
+	}, nil
+}
+
+// for local dir imports
+type fileSyncServerProxy struct {
+	c *Client
+}
+
+func (p *fileSyncServerProxy) Register(srv *grpc.Server) {
+	filesync.RegisterFileSyncServer(srv, p)
+}
+
+func (p *fileSyncServerProxy) DiffCopy(stream filesync.FileSync_DiffCopyServer) error {
+	ctx, opts, err := p.c.getLocalImportOpts(stream)
+	if err != nil {
+		return err
+	}
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	diffCopyClient, err := filesync.NewFileSyncClient(opts.session.Conn()).DiffCopy(ctx)
+	if err != nil {
+		return err
+	}
+	return proxyStream[filesynctypes.Packet](ctx, diffCopyClient, stream)
+}
+
+func (p *fileSyncServerProxy) TarStream(stream filesync.FileSync_TarStreamServer) error {
+	ctx, opts, err := p.c.getLocalImportOpts(stream)
+	if err != nil {
+		return err
+	}
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	tarStreamClient, err := filesync.NewFileSyncClient(opts.session.Conn()).TarStream(ctx)
+	if err != nil {
+		return err
+	}
+	return proxyStream[filesynctypes.Packet](ctx, tarStreamClient, stream)
+}
+
+// TODO: workaround needed until upstream fix: https://github.com/moby/buildkit/pull/4049
+func (c *Client) newFileSendServerProxySession(ctx context.Context, destPath string) (*bksession.Session, error) {
+	clientMetadata, err := engine.ClientMetadataFromContext(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get requester client metadata: %s", err)
+	}
+	sess, err := bksession.NewSession(ctx, identity.NewID(), "")
+	if err != nil {
+		return nil, err
+	}
+	proxy := &fileSendServerProxy{c: c, destClientID: clientMetadata.ClientID, destPath: destPath}
+	sess.Allow(proxy)
+
+	clientConn, serverConn := net.Pipe()
+	dialer := func(ctx context.Context, proto string, meta map[string][]string) (net.Conn, error) { // nolint: unparam
+		go func() {
+			defer serverConn.Close()
+			err := c.SessionManager.HandleConn(ctx, serverConn, meta)
+			if err != nil {
+				lg := bklog.G(ctx).WithError(err)
+				if errors.Is(err, context.Canceled) || errors.Is(err, io.EOF) {
+					lg.Debug("session conn ended")
+				} else {
+					// TODO: cancel the whole buildkit client
```

Comment:
(seems low prio, flagging anyway)

Reply:
Ah yeah, I meant to delete this comment, it turns out to not be needed since everything should error out how we'd want if this branch is hit.